presto-hive-common/src/main/java/com/facebook/presto/hive/zorder/ZOrderCurveSearcher.java

```diff
@@ -0,0 +1,233 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.facebook.presto.hive.zorder;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
+import java.util.stream.Collectors;
+
+import static com.google.common.base.Preconditions.checkArgument;
+import static java.lang.String.format;
+
+/**
+ * The ZOrderCurveSearcher class contains functions to search for addresses within given value ranges.
+ */
+public class ZOrderCurveSearcher
+{
+    public enum ReturnType {
+        INTEGER, LONG
+    }
+
+    private final ZOrder zOrder;
+
+    private final List<Integer> encodingBits;
+    private final int totalBitLength;
+    private final int maxBitLength;
+
+    private List<ZValueRange> ranges;
```

Comment:
Does this `ranges` indicate some state of the class? Can we avoid making the a non-final field of the class by passing this between class methods?